### PR TITLE
Add a basic support for the categorical type.

### PR DESCRIPTION
**Pull request by ueshin**
Experimental.

Add a basic support for the categorical type.

```py
>>> s = ks.Series(list("abbccc"), dtype="category")
>>> s
0    a
1    b
2    b
3    c
4    c
5    c
dtype: category
Categories (3, object): ['a', 'b', 'c']
>>> s.cat.categories
Index(['a', 'b', 'c'], dtype='object')
>>> s.cat.codes
0    0
1    1
2    1
3    2
4    2
5    2
dtype: int8

>>> idx = ks.CategoricalIndex(list("abbccc"))
>>> idx
CategoricalIndex(['a', 'b', 'b', 'c', 'c', 'c'],
                 categories=['a', 'b', 'c'], ordered=False, dtype='category')

>>> idx.codes
Int64Index([0, 1, 1, 2, 2, 2], dtype='int64')
>>> idx.categories
Index(['a', 'b', 'c'], dtype='object')
```

Currently type conversion with `astype` from/to the categorical type is supported, and other operations are not supported yet.

### CODE REVIEW

**Comment by ueshin** [comment]
This is based on #2061.

**Review by HyukjinKwon — COMMENTED** [review]
Looks fine from a cursory look

File: databricks/koalas/base.py [hunk 1]
```diff
@@ -1056,6 +1056,54 @@ def astype(self, dtype: Union[str, type, Dtype]) -> Union["Index", "Series"]:
...
+                codes, uniques = self.factorize()
+                return codes._with_new_scol(
+                    codes.spark.column, dtype=CategoricalDtype(categories=uniques)
+                )
+            else:
+                categories = dtype.categories
+                if len(categories) == 0:
+                    scol = F.lit(-1)
+                else:
+                    kvs = list(
```
Comment (HyukjinKwon) [inline databricks/koalas/base.py:1092]: Maybe we could think about having a util for this codes ...

Comment (ueshin) [reply databricks/koalas/base.py:1092]: Yep, I'm going to do some refactoring in the next PR.

**Comment by ueshin** [comment]
Thanks! Let me merge this now. Please feel free to leave comments.

**Pull request merged**